[feature] Add a check which inspects device configuration status periodically

nepython opened the pull request:
Closes #54

nepython commented:
**QUERIES**:
* When is the status supposed to be made critical? The issue metioned that status should change to `problem/critical` depending upon the configuration. I didn't understood this. 
* I think that I should rename all instances of config_status to config_modified, it looks more appropriate. Can I please know if it seems a better name?

PabloCastellano commented:
> **QUERIES**:
> 
> * When is the status supposed to be made critical? The issue metioned that status should change to `problem/critical` depending upon the configuration. I didn't understand this.

Check out the README file. According to the current version, a status will be critical when the metric is defined in `OPENWISP_MONITORING_CRITICAL_DEVICE_METRICS`. Critical status is reserved for some metrics such as ping.

PabloCastellano commented on README.rst:
```diff
@@ -247,6 +247,47 @@ in terms of disk space.
+| **default**: | ``True``    |
++--------------+-------------+
+
+Whether ``config_modified`` checks are created automatically for devices.
```
This should be explained in a better way, since `config_modified check` doesn't say much at all to newcomers

nepython replied:
I was trying to be consistent with the `ping` check but yes you are right. I shall work upon improving the documentation :smiley:

PabloCastellano commented on README.rst:
```diff
@@ -247,6 +247,47 @@ in terms of disk space.
+This fixed **duration** is 5 minutes by defaut and can be changed with the help of this setting.
+The input represents corresponding duration in minutes.
+
+**Note**: If the setting ``AUTO_CONFIG_STATUS`` is disabled then this setting need not be declared.
```
What is this AUTO_CONFIG_STATUS setting? Please review that everything makes sense when making changes and submitting PRs

PabloCastellano commented on openwisp_monitoring/check/settings.py:
```diff
@@ -3,7 +3,21 @@
+)
+# Input in days
+CONFIG_MODIFIED_RETENTION_POLICY = getattr(
+    settings, 'OPENWISP_MONITORING_CONFIG_STATUS_RETENTION_POLICY', '48h0m0s'
```
Please follow the standard we use for setting names: the user defines `OPENWISP_MONITORING_X_Y_Z` but we use `X_Y_Z` internally

nepython replied:
Can there be an exception here. I have named the setting as `OPENWISP_MONITORING_DEVICE_CONFIGURATION_X_Y_Z` but internally the settings are named as `CONFIG_X_Y_Z`. The reason was to explicitly say to newcomers that the setting is related to `DEVICE_CONFIGURATION` (there's also chart config and soon there shall be metric_config) but internally i.e. in the `check settings` we know that it means device config hence it's shortened. Let me know if it seems right?

PabloCastellano commented on openwisp_monitoring/check/tasks.py:
```diff
@@ -54,3 +54,32 @@ def auto_create_ping(model, app_label, object_id, created):
+            object_id=object_id,
+            content_type__model='device',
+            check=config_modified_path,
+        ).count()
```
Tip: there is `.exists()` which replaces `.count() > 0`

nepython replied:
I do know :sweat_smile:. I thought of using it but then dropped the idea to be consistent with `ping`'s code. Thanks, I shall change for both the checks, then :)

PabloCastellano commented on openwisp_monitoring/check/base/models.py:
```diff
@@ -75,27 +76,45 @@ def perform_check(self, store=True):
         return self.check_instance.check(store=True)
 
 
-if app_settings.AUTO_PING:
```
I preferred the previous approach to be honest because previously we only registered the dispatcher when necessary (AUTO_PING=True). Now it is always registered (no matters the value of AUTO_PING) and hence executed on each post_save signal, causing the function to do nothing when AUTO_PING=False, note the difference.

nepython replied:
Makes sense but then I shall have to add a check in corresponding tasks. The reason being, the dispatcher shall get registered when the models are loaded. So, there is no option of mocking `AUTO_PING`, `AUTO_CONFIG_STATUS` it seems. Except for the tests, there is no real need of the check in the tasks.

PabloCastellano commented on README.rst:
```diff
@@ -247,6 +247,47 @@ in terms of disk space.
+
+After ``config`` is ``modified``, if the ``modified`` status does not change  after a
+fixed **duration** then ``device`` health status shall change to ``problem``.
+This fixed **duration** is 5 minutes by defaut and can be changed with the help of this setting.
```
typo: defaut

PabloCastellano commented on README.rst:
```diff
@@ -247,6 +247,47 @@ in terms of disk space.
+
+Whether ``config_modified`` checks are created automatically for devices.
+
+``OPENWISP_MONITORING_CONFIG_MODIFIED_MAX_TIME``
```
Incorrect: you have defined `OPENWISP_MONITORING_CONFIG_MODIFIED_MAXIMUM_TIME` in settings.py

PabloCastellano commented on README.rst:
```diff
@@ -247,6 +247,47 @@ in terms of disk space.
+| **default**: |   ``5``   |
++--------------+-----------+
+
+After ``config`` is ``modified``, if the ``modified`` status does not change  after a
```
Sounds very technical too.

What about something like

> This setting allows configuring the time that the check is allowed to fail until the device health status changes to ``problem``

It is important to take the necessary time to re-read the documentation multiple times to make sure it is written properly and others can understand it. It is not easy though!

PabloCastellano commented on openwisp_monitoring/check/tasks.py:
```diff
@@ -54,3 +54,32 @@ def auto_create_ping(model, app_label, object_id, created):
+        return
+    ct = ContentType.objects.get(app_label=app_label, model=model)
+    check = Check(
+        name='Config Modified',
```
What about 'Device configuration' or 'Configuration Modified'? The name shouldn't contain abbreviations

nepython commented:
@nemesisdesign, I realized it now that there is no migration for `ping`, `config_modified` checks in the `sample_check`. I don't think there's a need since the module is not yet released (and the migrations were meant for backward compatibilty). Can you please confirm this?

nepython commented on openwisp_monitoring/check/tasks.py:
```diff
@@ -35,22 +35,43 @@ def perform_check(uuid):
 
 @shared_task
-def auto_create_ping(model, app_label, object_id, created):
+def auto_create_ping(model, app_label, object_id):
```
I think it's safe to remove the the keyword argument `created` , since it's not being used at all.

nemesisdesign commented on README.rst:
```diff
@@ -247,6 +247,53 @@ in terms of disk space.
+
+``config_modified`` checks are run peridically to monitor whether device
+configuration is applied properly.
+This setting is used to govern whether ``config_modified`` checks are
```
govern? Can you find a different way of explaining this?

nemesisdesign commented on README.rst:
```diff
@@ -247,6 +247,53 @@ in terms of disk space.
 
 Whether ping checks are created automatically for devices.
 
+``OPENWISP_MONITORING_AUTO_DEVICE_CONFIGURATION_MODIFIED``
```
what about `OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK`?

nemesisdesign commented on README.rst:
```diff
@@ -247,6 +247,53 @@ in terms of disk space.
++--------------+-----------+
+
+This setting allows you to configure the time that the check is allowed to
+fail after which the device health status changes to ``problem``.
```
I don't understand this setting from its explanation, we should also find a shorter name, eg `OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME`

nemesisdesign commented on README.rst:
```diff
@@ -247,6 +247,53 @@ in terms of disk space.
+**Note**: If the setting ``OPENWISP_MONITORING_AUTO_DEVICE_CONFIGURATION_MODIFIED``
+is disabled then this setting need not be declared.
+
+``OPENWISP_MONITORING_DEVICE_CONFIGURATION_MODIFIED_RETENTION_POLICY``
```
too long.. `OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_RP`?

nemesisdesign commented on openwisp_monitoring/check/apps.py:
```diff
@@ -1,8 +1,13 @@
     verbose_name = _('Network Monitoring Checks')
+
+    def ready(self):
+        manage_config_modified_retention_policy()
```
how does this play with the abstraction? Can you make it play well with the changes introduced in #103?

nepython replied:
`manage_config_modified_retention_policy()` makes use of `timeseries_db.get_list_retention_policies` after rebasing which has been defined by us so it's abstract enough, I think.

nemesisdesign commented on openwisp_monitoring/check/classes/base.py:
```diff
@@ -0,0 +1,38 @@
+Metric = load_model('monitoring', 'Metric')
+
+
+class BaseCheck(object):
```
I would add:

```python
def validate(self):
    self.validate_instance()
    self.validate_params()

def validate_params(self):
    pass
```

nepython replied:
Thanks!

nemesisdesign commented on README.rst:
```diff
@@ -247,6 +247,53 @@ in terms of disk space.
+
+**Note**: If the setting ``OPENWISP_MONITORING_AUTO_DEVICE_CONFIGURATION_MODIFIED``
+is disabled then this setting need not be declared.
+
```
I think we should start creating a section which describes the available checks, which can go above, for now we'll have two checks: ping and this new one, we need a description for each one, and each section can then link the available settings for each check.

I would like to see a prose description of this new check.

nemesisdesign commented on openwisp_monitoring/check/classes/base.py:
```diff
@@ -0,0 +1,45 @@
+
+    def validate_params(self):
+        pass
+
```
I forgot something else, which is just something which is usually done when developing these kind of object oriented patterns:

```python
def check(self, store=True):
    raise NotImplementedError
```

It's usage is recommended in the python docs, I suggest to read this passage: https://docs.python.org/3/library/exceptions.html#NotImplementedError

PabloCastellano commented on README.rst:
```diff
@@ -224,6 +224,32 @@ has a value which is not in the expected range
+``Ping``
+~~~~~~~~
+
+This check returns information on device ``uptime``, ``downtime``,
```
downtime is redundant, saying uptime is enough

PabloCastellano commented on README.rst:
```diff
@@ -224,6 +224,32 @@ has a value which is not in the expected range
+This check returns information on device ``uptime``, ``downtime``,
+``RTT (Round trip time) maximum, minimum and average``.
+It creates a ``Ping`` metric to store this information and an ``AlertSettings``
+to monitor them.
```
> It creates a ``Ping`` metric to store this information and an ``AlertSettings`` to monitor them.

These are technical details that a normal user prefers not to know

PabloCastellano commented on README.rst:
```diff
@@ -224,6 +224,32 @@ has a value which is not in the expected range
+to monitor them.
+The Charts ``uptime``, ``packet loss`` and ``rtt`` are created. The ``fping``
+command is used to collect these metrics.
+You may choose to disable auto creation of this check by using the setting
```
Instead of `by using the setting ...` I'd say `by setting ... to False` which is more concrete

PabloCastellano commented on README.rst:
```diff
@@ -224,6 +224,32 @@ has a value which is not in the expected range
+the device health status to ``PROBLEM``, if config status stays modified
+for more than `OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME <#OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME>`_.
+You may choose to disable auto creation of this check by using the
+setting `OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK <#OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK>`_.
```
Same: say "setting it to False"

PabloCastellano commented on README.rst:
```diff
@@ -224,6 +224,32 @@ has a value which is not in the expected range
+``Configuration Modified``
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It checks the configuration status of the device periodically and changes
```
What about this rewording

"This will periodically check the configuration present in the device and will change the device health status to ``PROBLEM`` if the configuration is not in sync with the expected by OpenWISP. The time until this check changes the health status can be configured using the setting `OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME`"

nepython replied:
It sounds a bit technical.

PabloCastellano commented on README.rst:
```diff
@@ -253,6 +279,49 @@ in terms of disk space.
+``OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK``. It's 5 minutes by default.
+
+**Note**: If the setting ``OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK``
+is disabled then this setting need not be declared.
```
"This setting will be ignored if `OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK = False`"

PabloCastellano commented on README.rst:
```diff
@@ -253,6 +279,49 @@ in terms of disk space.
+
+``config_modified`` check changes the device health status to ``PROBLEM`` if it's
+configuration status has been in the modified state for more than  the time defined in
+``OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK``. It's 5 minutes by default.
```
"Defines the amount of time that the device configuration can be different from what is expected until its health status is changed to PROBLEM"

nemesisdesign replied:
```
Defines the maximum amount of minutes the device configuration can stay in the *modified* state before its health status is changed to PROBLEM and an alert is sent.
```

PabloCastellano commented on README.rst:
```diff
@@ -253,6 +279,49 @@ in terms of disk space.
++--------------+-------------+
+| **type**:    |  ``time``   |
++--------------+-------------+
+| **default**: | ``48h0m0s`` |
```
I think this is actually too low. Can you set it to 30 days by default?

nepython replied:
I can set it but the issue (derived from measurable outcomes) mentioned a default retention period of 2 days. It made sense since I don't think the user would be interested to know this info at all and it's for internal usage.

PabloCastellano commented on openwisp_monitoring/check/classes/base.py:
```diff
@@ -0,0 +1,48 @@
+Metric = load_model('monitoring', 'Metric')
+
+
+class BaseCheck(object):
```
While refactoring BaseCheck is needed, this kind of changes make it more difficult to review PRs and detect problems later. It is usually a good thing to make one change per PR so please consider it next time (1 PR for refactoring and another one to add this new check).

PabloCastellano commented on openwisp_monitoring/check/classes/base.py:
```diff
@@ -0,0 +1,48 @@
+            field_name=field_name,
+            key=self.__class__.__name__.lower(),
+        )
+        metric, created = Metric.objects.get_or_create(**options)
```
Shouldn't we use `Metric._get_or_create` ?

nemesisdesign commented on openwisp_monitoring/device/apps.py:
```diff
@@ -162,3 +164,23 @@ def register_notifcation_types(self):
+
+        DeviceData = load_model('device_monitoring', 'DeviceData')
+        device = DeviceData.objects.get(config=instance)
+        check = device.checks.get(name='Configuration Modified')
```
never query in the code by name, it's too fragile!

The name right now can also be changed by users so this would stop working in that case.
Use another field, eg: `configuration`.

Moreover, what if the Check has been deleted? Please write a test for this case and ensure it's handled without crashes.

nepython commented:
> What if the device is already in critical state (eg: turned off)?
> 
> The scenario is the following.
> 
> User has 10 devices using the same configuration template.
> 3 devices are offline (unplugged, turned off, whatever).
> User edits one of the common templates, hence the configuration status of all the devices changes to modified.
> The 3 devices that are offline will not get their config status updated because they're offline.
> 
> The user does not expect (nor wants) to not be notified about this.
> The user does expect to know if there's any issue once the devices do come back online.

How's this. If the device health status becomes `critical` (due to any reason), config modified checks won't be run thus there won't be metric write, threshold_crossed and thus no alert generated.
When the device comes back online (not critical anymore), `config_modified` checks shall be triggered and if status stays `modified` for more than threshold period, an alert will be sent.
Thus besides the celery tasks, `config_modified` checks shall be triggered when config gets modified / device comes back online.

This should cover the case, so let me know :smile:

nepython commented:
> Is this ready for review/testing?

Can you please confirm that https://github.com/openwisp/openwisp-monitoring/pull/123#issuecomment-654134969 covers the use case well, since I am unsure about it.

nepython commented:
@nemesisdesign, I went ahead with https://github.com/openwisp/openwisp-monitoring/pull/123#issuecomment-654134969, this is the code that I have added
https://github.com/openwisp/openwisp-monitoring/blob/bf712870125a05d347436857ccc7fc52289bae16/openwisp_monitoring/check/classes/config_modified.py#L12-L14
It checks device status, if it's `critical` then `config_modified` checks won't be run and thus no corresponding alerts. I think it covers the use case well :smile:.

nemesisdesign commented on README.rst:
```diff
@@ -242,6 +251,33 @@ in terms of disk space.
+| **default**: | ``True``    |
++--------------+-------------+
+
+This setting allows you to choose whether `config_modified <#Configuration>` checks should be
```
see how it looks: https://github.com/openwisp/openwisp-monitoring/blob/125f1c7fb11280cb3435f6545cf7171b893d9aaa/README.rst#id2

Always do these manual visual checks before publishing docs.

nemesisdesign commented on openwisp_monitoring/check/base/models.py:
```diff
@@ -91,3 +94,22 @@ def auto_ping_receiver(sender, instance, created, **kwargs):
         )
+
+
+def auto_config_modified_receiver(sender, instance, created, **kwargs):
```
instead of referring to `config_modified`, let's refer to `config_check`, follow the same suggestion for all other occurrences

nemesisdesign commented on openwisp_monitoring/monitoring/metrics.py:
```diff
@@ -12,6 +12,12 @@
         'field_name': 'reachable',
         'related_fields': ['loss', 'rtt_min', 'rtt_max', 'rtt_avg'],
     },
+    'configmodified': {
```
here I would use `config_applied` and same for `key`and `field_name`

nepython commented:
> Thank you @nepython, I tested it and it works! +1
> 
> After using the feature and analyzing the code, I realized that the naming is odd.
> 
> The check would be a lot more easy to understand if it's something like:
> 
> name: Configuration applied
> alert settings operator: less than
> alert settings threshold value: 1
> alert settings threshold value: 600
> 
> So if the configuration is not applied for more than 5 minutes, we get an alert.
> 
> PS: while testing this, I also realized this: #172

@nemesisdesign, I am slightly confused. You are saying it should be 5 min and 600s at the same time :thinking:.
Currently it defaults to 5 minutes with a setting `OPENWISP_MONITORING_CONFIG_CHECK_MAX_TIME` to allow user to configure if needed (as per the issue description). I think the current approach of default 5 min with a setting to configure is better. Wdyt?

nemesisdesign commented on openwisp_monitoring/check/migrations/0004_create_config_applied.py:
```diff
@@ -0,0 +1,33 @@
+
+def remove_config_applied_checks(apps, schema_editor):
+    Check = apps.get_model('config', 'Device')
+    Check.objects.filter(name='Configuration Applied').delete()
```
you should never rely on fields that can be changed like names and descriptions, you should query based on data which  cannot be altered by users.

nemesisdesign commented on openwisp_monitoring/check/migrations/0004_create_config_applied.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def remove_config_applied_checks(apps, schema_editor):
+    Check = apps.get_model('config', 'Device')
```
**TOTALLY BOGUS**. 19 days should have been enough to test this and make it work but yet, you fail at this Hardik. :-1: you're still sending broken code for final review after all the discussions we had. You have to stop doing this.

nemesisdesign replied:
1. try to run this
2. make sure it deletes also the Metric (and related alert setting cascades)

nemesisdesign commented on README.rst:
```diff
@@ -213,6 +213,15 @@ command is used to collect these metrics.
+the device health status to ``PROBLEM``, if config status stays modified
+for more than `OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME <#OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME>`_.
+You may choose to disable auto creation of this check by using the
+setting `OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK <#OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK>`_.
```
I think this is better:

```
``Configuration applied``
~~~~~~~~~~~~~~~~~~~~~~~~~

This check ensures that the `openwisp-config agent
<https://github.com/openwisp/openwisp-config/>`_ is running and applying
configuration changes in a timely manner.

By default, if a configuration object stays in the ``modified`` state
for more than 5 minutes (configurable via
`OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME
<#OPENWISP_MONITORING_DEVICE_CONFIG_CHECK_MAX_TIME>`_), the check will
fail and set the health status of the device to ``PROBLEM``.

You may choose to disable auto creation of this check by using the
setting `OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK
<#OPENWISP_MONITORING_AUTO_DEVICE_CONFIG_CHECK>`_.
```

nemesisdesign commented on openwisp_monitoring/check/tests/test_models.py:
```diff
@@ -42,34 +56,128 @@ def test_base_check_class(self):
+            self.assertEqual(c1.content_object, d)
+            self.assertIn('Ping', c1.check)
+        with self.subTest('Test AUTO_CONFIG_CHECK'):
+            c2 = Check.objects.get(name='Configuration Applied')
```
don't query by name, use `Check.objects.filter(check=self._CONFIG_APPLIED).first()`

fix the same issue for all other occurrences across the codebase!

nemesisdesign commented on openwisp_monitoring/check/tests/test_models.py:
```diff
@@ -42,34 +56,128 @@ def test_base_check_class(self):
-        self.assertIn('Ping', c.check)
+        self.assertEqual(Check.objects.count(), 2)
+        with self.subTest('Test AUTO_PING'):
+            c1 = Check.objects.get(name='Ping')
```
don't query by name, use the check field, please correct the same for any other occurrence of this everywhere!
Do not query by fields that can be changed by users, do not use get() for querying objects that can return multiple items!

nepython replied:
> do not use get() for querying objects that can return multiple items!

I expect one `Ping` and one `ConfigApplied` check per device. Since, the test had only one device, I thought using `get` was safe enough. Though yes, querying by `name` (which can be changed) is bad practice on my side :(

nemesisdesign commented on openwisp_monitoring/check/tests/test_models.py:
```diff
@@ -42,34 +56,128 @@ def test_base_check_class(self):
+        with self.subTest('Test AUTO_CONFIG_CHECK'):
+            c2 = Check.objects.get(name='Configuration Applied')
+            self.assertEqual(c2.content_object, d)
+            self.assertIn('ConfigApplied', c2.check)
```
why don't you use equal and you compare it to `self._CONFIG_APPLIED`

nepython replied:
This test was already written for `Ping` check. So, I have just adapted it for `ConfigApplied` checks. Made the change.

nemesisdesign commented on openwisp_monitoring/check/classes/config_applied.py:
```diff
@@ -0,0 +1,35 @@
+
+    def _create_alert_setting(self, metric):
+        alert_s = AlertSettings(
+            metric=metric, operator='<', value=1, seconds=CONFIG_CHECK_MAX_TIME * 60
```
keep in mind this will have to be changed when we do https://github.com/openwisp/openwisp-monitoring/issues/106

nepython commented:
> Also, the check doesn't seem to work to me if the tolerance is not zero seconds, I'll do more tests and let you know more details if I can. **UPDATE**: I confirm this, here's how I test it:
> 
>     * run dev server, run worker, run celery beat
> 
>     * set the config of a device to `modified`
> 
>     * wait more than 5 minutes, no alert is ever generated
> 
>     * if I set `seconds` to `0` in the alert settings, the alert is generated

@nemesisdesign, no it's not a bug. You had asked to cover this [use case](https://github.com/openwisp/openwisp-monitoring/pull/123#pullrequestreview-442628533) earlier :confused:, I replied back with a solution https://github.com/openwisp/openwisp-monitoring/pull/123#issuecomment-654134969. Awaited your response, pinged you two days later https://github.com/openwisp/openwisp-monitoring/pull/123#issuecomment-655560372 and on IM. Still no response, maybe you were slightly busy back then. Hence, went ahead with the approach and informed nevertheless https://github.com/openwisp/openwisp-monitoring/pull/123#issuecomment-656330919.

As you had requested, the `ConfigApplied` check won't send any alerts if the device goes offline (health is thus `critical`). The user will already have one for `Ping`. The reason being to avoid irritating the user with too many alerts. It will do alert the user once device comes back online and `config` still stays modified.

The zero second case I think is an edge case where device goes offline. `ConfigApplied` checks are triggered prior to health status getting updated to `CRITICAL`. We might never set a `0s` threshold for `ConfigApplied` I think, since config status stays modified for some time whenever a template is updated (it's not spontaneous atleast for me :sweat_smile:)

**Note**: There is one specific test (`test_config_modified_device_problem`) that I have added in `check.tests.TestModels` which tests that the `ConfigApplied` check is working!

nemesisdesign commented on openwisp_monitoring/check/classes/config_applied.py:
```diff
@@ -0,0 +1,35 @@
+            return
+        if not hasattr(self.related_object, 'config'):
+            return
+        result = 1 if self.related_object.config.status == 'applied' else 0
```
this could be rewritten as:

```python
int(self.related_object.config.status == 'applied')
```

not sure which one is more readable, I think this form is compact and readable

nepython commented:
Mentioning so if anyone faces same issue in future, then can avoid the trouble, hopefully.
> Please repeat the same tests and let me know:
> 
>     1. if you can replicate the same issue successfully

Yes
>     2. if yes, why it is happening? Why writing with the short retention policy doesn't write anything?

When a retention policy is used besides `default` in `influxdb` then it needs to be specified during query.
https://community.influxdata.com/t/influxdb-not-writing-data-if-rp-is-specified/9861. So, the issue is not with `write` but with `read`. To query any measurement in `influxdb` with a non-default retention_policy in place, please use.
```
> select * from <retention_policy>.<measurement>;   # `select * from short.config_applied` in this case 
```
>     3. if yes, write a failing test for it

Added `openwisp_monitoring.db.backends.influxdb.tests.TestDatabaseClient.test_read_with_rp`
>     4. if yes, how you fixed it

Added `retention_policy` as a parameter in `metric.read()` and `timeseries_db.read()`. So, you can query like this,
```python
In [7]: m.read(retention_policy='short')
Out[7]: 
[{'time': 1596102060, 'config_applied': 1},
 {'time': 1596102120, 'config_applied': 1},
...
```
Personally, I don't like this approach since user will need to specify the `retention_policy` everytime in read / write. Maybe we can refactor `retention_policy` to be a field in `Metric` model (later on).

The alert didn't came immediately after 5 minutes (took roughly 6-7 minutes, I think) but yes tested twice and added `Notification count` in  `test_config_modified_device_problem`
![Screenshot from 2020-07-31 02-15-21](https://user-images.githubusercontent.com/54471024/88978187-d962f280-d2dc-11ea-8bf9-8dd1cb9418fa.png)

nemesisdesign commented on openwisp_monitoring/db/backends/influxdb/client.py:
```diff
@@ -130,11 +130,15 @@ def read(self, key, fields, tags, **kwargs):
+        if not retention_policy:
+            q = f'SELECT {fields} FROM {key}'
+        else:
+            q = f'SELECT {fields} FROM {retention_policy}.{key}'
```
do not repeat the whole query, follow the DRY principle please:

```python
from_clause = key
if retention_policy:
    from_clause = f'{retention_policy}.{key}'
q = f'SELECT {fields} FROM {from_clause}'
```

or a more compact version:

```python
from_clause = f'{rt}.{key}' if rt else key
q = f'SELECT {fields} FROM {from_clause}'
```

nepython commented:
> > > ```
> > > 4. if yes, how you fixed it
> > > ```
> > 
> > 
> > Added `retention_policy` as a parameter in `metric.read()` and `timeseries_db.read()`.
> 
> Ok, I see you added support for this in the tsdb backend, but the short retention policy was already being used to retrieve the device data, and that code has not been updated. Can you update that bit of code as well to make it use the new way so we can probably simplify the code? I'm referring to these:
> 
> https://github.com/openwisp/openwisp-monitoring/blob/87e7135b4b6e0b4a0c744f0aaa87b021912eae50/openwisp_monitoring/device/base/models.py#L99
> 
> https://github.com/openwisp/openwisp-monitoring/blob/238a17bcfebef619c5c125a130abcb076d1a9c4b/openwisp_monitoring/db/backends/influxdb/queries.py#L67

@nemesisdesign, as you have correctly pointed it out, `timseries_db.query()` was being used and not `timeseries_db.read()` for `device_data`. This method of how we query `device_data` will change fully after #164 (reason is elasticsearch doesn't support querying large text fields (efficiency issues) so had to adopt a different approach). It will simplify and transform as below,
https://github.com/openwisp/openwisp-monitoring/blob/9c53057826befd121d92035885a972163b93134c/openwisp_monitoring/device/base/models.py#L100-L102
also abstraction was not complete as `device_data_query` is a `dict` in case of `elasticsearch`. I think if needed, we can simplify it there :smiley:.

> In order to have robust tests, I think we should switch to use freezegun to arbitrarily manipulate the time, we already do this in other modules, can you please take advantage of this chance to change all the tests of this test suite which have this approach so that they use freezegun.freeze_time instead?
> 
> You would freeze the time 10 minutes earlier to trigger the check, that would trigger the insertion in the timeseries DB without the need of this line, this will make the testcases a lot more robust and will help us catch bugs in the future.

Ok, thanks freezegun works great. Currently, implemented only for this test, will do for all others in the suite. Though we will need to run checks twice in order to validate alerts after this. The reason being `datetime.now()` is then mocked throughout the code to point 10 minutes backwards. Hence, `alert_settings._is_crossed` also gets `now()` as `now() - timedelta(minutes=10)`. So, two times `perform_check` needs to be run, one with `freeze_time` (context manager) and one without it. It works perfectly with this.

To get it working, I had to alter tsdb backend `write` method, this seems to be causing failure of many other tests (maybe clock differences (unsure)). I am checking out why, change is simple enough :thinking:.
```python
timestamp = kwargs.get('timestamp') or now()  # <--- now() is extra here
if isinstance(timestamp, datetime):
    timestamp = timestamp.strftime('%Y-%m-%dT%H:%M:%SZ')
point['time'] = timestamp
```

nepython commented:
> To get it working, I had to alter tsdb backend write method, this seems to be causing failure of many other tests (maybe clock differences (unsure)). I am checking out why, change is simple enough thinking.

Check this build (https://travis-ci.org/github/openwisp/openwisp-monitoring/jobs/713662695#L1163) to know the failure.

Found it, it was related to `precision` of the timestamp. This issue was hidden till now and came forward now due to below change.
```python
timestamp = kwargs.get('timestamp') or now()  # now() is added here
if isinstance(timestamp, datetime):
    timestamp = timestamp.strftime('%Y-%m-%dT%H:%M:%SZ')  # This was the reason, seconds precision
point['time'] = timestamp    # Passing timestamp was optional earlier
```
I will try to add a failing test for this.
